Objects/typevarobject.c

```diff
@@ -0,0 +1,1724 @@
+                 PyObject *bound, int covariant, int contravariant,
+                 int infer_variance)
+/*[clinic end generated code: output=463d42203b82ea59 input=55a3db20c1cf1450]*/
+{
+    if (covariant && contravariant) {
+        PyErr_SetString(PyExc_ValueError,
+                        "Bivariant types are not supported.");
+        return NULL;
+    }
+
+    if (infer_variance && (covariant || contravariant)) {
+        PyErr_SetString(PyExc_ValueError,
+                        "Variance cannot be specified with infer_variance.");
+        return NULL;
+    }
+
+    if (Py_IsNone(bound)) {
+        bound = NULL;
+    }
+    if (bound != NULL) {
+        bound = type_check(bound);
+        if (bound == NULL) {
+            return NULL;
+        }
+    }
+
+    if (!PyTuple_CheckExact(constraints)) {
+        PyErr_SetString(PyExc_TypeError,
+                        "constraints must be a tuple");
+        return NULL;
+    }
+    Py_ssize_t n_constraints = PyTuple_GET_SIZE(constraints);
+    if (n_constraints == 1) {
+        PyErr_SetString(PyExc_TypeError,
+                        "A single constraint is not allowed");
+        Py_XDECREF(bound);
+        return NULL;
+    } else if (n_constraints == 0) {
+        constraints = NULL;
+    } else if (bound != NULL) {
+        PyErr_SetString(PyExc_TypeError,
+                        "Constraints cannot be combined with bound=...");
+        Py_XDECREF(bound);
+        return NULL;
+    }
+    PyObject *module = caller();
+    if (module == NULL) {
+        Py_XDECREF(bound);
+        return NULL;
+    }
+
+    PyObject *tv = (PyObject *)typevar_alloc(name, bound, NULL,
+                                             constraints, NULL,
+                                             covariant, contravariant,
+                                             infer_variance, module);
+    Py_XDECREF(bound);
+    Py_XDECREF(module);
+    return tv;
+}
+
+/*[clinic input]
+typevar.__typing_subst__ as typevar_typing_subst
+
+    arg: object
+
+[clinic start generated code]*/
+
+static PyObject *
+typevar_typing_subst_impl(typevarobject *self, PyObject *arg)
+/*[clinic end generated code: output=c76ced134ed8f4e1 input=6b70a4bb2da838de]*/
+{
+    PyObject *args = PyTuple_Pack(2, self, arg);
+    if (args == NULL) {
+        return NULL;
+    }
+    PyObject *result = call_typing_func_object("_typevar_subst", args);
+    Py_DECREF(args);
+    return result;
+}
+
+/*[clinic input]
+typevar.__reduce__ as typevar_reduce
+
+[clinic start generated code]*/
+
+static PyObject *
+typevar_reduce_impl(typevarobject *self)
+/*[clinic end generated code: output=02e5c55d7cf8a08f input=de76bc95f04fb9ff]*/
+{
+    return PyUnicode_FromString(self->name);
+}
+
+static PyObject *
+typevar_mro_entries(PyObject *self, PyObject *args)
+{
+    PyErr_SetString(PyExc_TypeError,
+                    "Cannot subclass an instance of TypeVar");
+    return NULL;
+}
```

Comment:
It mirrors the existing CPython code: https://github.com/python/cpython/blob/82f789be3b15df5f6660f5fd0c563ad690ee00fb/Lib/typing.py#L966. I think we added that because otherwise the error message is hard to understand when you try to do it.

Reply:
The method was added in https://github.com/python/cpython/pull/31148. Tbh, I'm not 100% convinced that the change was necessary.